core/pkg/artifacts/saver.go

```diff
@@ -301,6 +328,119 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
+	S3MaxParts           = 10000
+)
+
+func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+
+	if fileSize < S3MinMultiUploadSize {
+		// We don't need to use multipart for small files.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(0)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(path string, fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput, doneChan chan<- *filetransfer.Task,
+) {
+	statInfo, err := os.Stat(path)
+	if err != nil {
+		doneChan <- &filetransfer.Task{Err: err}
+		return
+	}
+	chunkSize := getChunkSize(statInfo.Size())
+
+	wg := sync.WaitGroup{}
+	subChan := make(chan *filetransfer.Task)
+	// TODO: add mid-upload cancel.
+
+	partInfo := fileInfo.MultipartUploadInfo
+	for i, part := range partInfo {
+		task := newUploadTask(fileInfo, path)
+		task.Offset = part.PartNumber * chunkSize
+		remainingSize := statInfo.Size() - task.Offset
+		if remainingSize < chunkSize {
+			task.Size = remainingSize
+		} else {
+			task.Size = chunkSize
+		}
+		task.Headers = append(task.Headers, "content-md5:"+partData[i].HexMD5)
+		task.Headers = append(task.Headers, "content-length:"+strconv.FormatInt(task.Size, 10))
+		task.SetCompletionCallback(func(t *filetransfer.Task) {
+			subChan <- t
+			wg.Done()
+		})
+		wg.Add(1)
+		as.FileTransferManager.AddTask(task)
+	}
+
+	go func() {
+		wg.Wait()
+		close(subChan)
+	}()
+
+	for t := range subChan {
+		if t.Err != nil {
+			doneChan <- t
+			return
+		}
+	}
+	_, err = gql.CompleteMultipartUploadArtifact(
+		as.Ctx, as.GraphqlClient, gql.CompleteMultipartActionComplete, partData,
+		fileInfo.BirthArtifactID, *fileInfo.StoragePath, fileInfo.UploadID,
+	)
+	task := newUploadTask(fileInfo, path)
+	task.Err = err
+	doneChan <- task
+}
+
+func getChunkSize(fileSize int64) int64 {
+	// Default to 100MiB chunks
+	const defaultChunkSize = int64(100 * 1024 * 1024)
+	if fileSize < defaultChunkSize*S3MaxParts {
+		return defaultChunkSize
+	}
+	// Use a larger chunk size if we would need more than 10,000 chunks.
+	chunkSize := int64(math.Ceil(float64(fileSize) / float64(S3MaxParts)))
+	chunkSize = int64(math.Ceil(float64(chunkSize)/4096) * 4096) // Round up to the nearest multiple of 4096.
```

Comment:
super nit: put the comment above the line to make it easier to read if one is reading just the comments and ignoring code.

Eyes have to scan the line to find the comment (also on my screen I have to scroll horizontally)
```
    // Use a larger chunk size if we would need more than 10,000 chunks.
    <code>
    <code>                                                       // Round up to the nearest multiple of 4096.
```

The comment starts in a natural spot:
```
    // Use a larger chunk size if we would need more than 10,000 chunks.
    <code>
    // Round up to the nearest multiple of 4096.
    <code>
```